src/Mod/PartDesign/App/FeatureRefine.cpp

```diff
@@ -0,0 +1,125 @@
+ *   License, or (at your option) any later version.                        *
+ *                                                                          *
+ *   FreeCAD is distributed in the hope that it will be useful, but         *
+ *   WITHOUT ANY WARRANTY; without even the implied warranty of             *
+ *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU       *
+ *   Lesser General Public License for more details.                        *
+ *                                                                          *
+ *   You should have received a copy of the GNU Lesser General Public       *
+ *   License along with FreeCAD. If not, see                                *
+ *   <https://www.gnu.org/licenses/>.                                       *
+ *                                                                          *
+ ***************************************************************************/
+
+
+#include "PreCompiled.h"
+#ifndef _PreComp_
+# include <Standard_Failure.hxx>
+#endif
+
+#include <App/Application.h>
+#include <App/FeaturePythonPyImp.h>
+#include <Base/Parameter.h>
+#include <Mod/Part/App/modelRefine.h>
+
+#include "FeatureRefine.h"
+#include "FeaturePy.h"
+
+
+using namespace PartDesign;
+
+namespace PartDesign
+{
+PROPERTY_SOURCE(PartDesign::FeatureRefine, PartDesign::Feature)
+
+FeatureRefine::FeatureRefine()
+{
+    ADD_PROPERTY_TYPE(Refine,(0),"Part Design",(App::PropertyType)(App::Prop_None),"Refine shape (clean up redundant edges) after operations");
+    //init Refine property
+    Base::Reference<ParameterGrp> hGrp = App::GetApplication().GetUserParameter()
+        .GetGroup("BaseApp")->GetGroup("Preferences")->GetGroup("Mod/PartDesign");
+    this->Refine.setValue(hGrp->GetBool("RefineModel", true));
+}
+
+bool FeatureRefine::onlyHasToRefine() const
+{
+    if( ! Refine.isTouched()){
+        return false;
+    }
+    if (rawShape.isNull()){
+        return false;
+    }
+    std::vector<App::Property*> propList;
+    getPropertyList(propList);
+    for (auto prop : propList){
+        if (prop != &Refine
+            /*&& prop != &SuppressedShape*/
+            && prop->isTouched()){
+            return false;
+        }
+    }
+    return true;
+}
+
+
+bool FeatureRefine::onlyHaveRefined()
+{
+    if (onlyHasToRefine()){
+        TopoShape result = refineShapeIfActive(rawShape);
+        Shape.setValue(result);
+        return true;
+    }
+    return false;
+}
+
+TopoShape FeatureRefine::refineShapeIfActive(const TopoShape& oldShape, const RefineErrorPolicy onError) const
+{
+    if (this->Refine.getValue()) {
+        TopoShape shape(oldShape);
+        try {
+            return shape.makeElementRefine();
+        }
+        catch (Standard_Failure& err) {
+            if(onError == RefineErrorPolicy::Warn){
+                Base::Console().Warning((std::string("Refine failed: ") + err.GetMessageString()).c_str());
```

Comment:
Warning does not support formatting patterns by default? This should use fmt or something like that.

Reply:
Switched to fmt::format